FAQ Tweaks, Including Root Chain FAQ + Terminology (#661)
cert-manager/website: content/en/docs/faq/terminology.md

---
title: "TLS Terminology"
linkTitle: "TLS Terminology"
weight: 50
type: "docs"
---

With TLS being so widely deployed, terminology can sometimes get confused or be used to mean different things, and that reality
combined with the complexity of TLS can lead to serious misunderstandings and confusion.

> 

Reviewer: Suggestion: we already linked it elsewhere, but maybe worth to link the TLS spec here too? https://datatracker.ietf.org/doc/html/rfc5246

### What does "publicly trusted" mean?

> Broadly speaking, a "publicly trusted" certificate is one that you can use on the public internet and expect
> that most reasonably up-to-date computers will be able to verify it using their system trust store.

Reviewer 2: ```suggestion
Broadly speaking, a "publicly trusted" certificate is one that you can use on the Internet and expect
that most reasonably up-to-date computers will be able to verify it using their system trust store.
```

Author: haha I'd not really thought about it but I guess "public internet" might be a bit redundant! thanks!

There isn't a single standard trust store containing certs which are "publicly trusted", but generally most

> of the commonly seen trust stores are similar. An example would be [Mozilla's CA Certificate Program](https://wiki.mozilla.org/CA).

Reviewer 2: I looked to see how Python certifi library get the trusted CAs:

https://github.com/certifi/python-certifi/blob/8110050d8441d363c8057d5ad187c8ad5a862fc2/Makefile#L1-L2

They use https://mkcert.org/ which might be worth linking to?

Reviewer 2: Which downloads trusted CAs from https://hg.mozilla.org/mozilla-central/raw-file/tip/security/nss/lib/ckfw/builtins/certdata.txt I think.

Author: My initial reaction here is that this might be out of scope for a terminology guide, because it's getting into pretty esoteric practices of customising trust stores. Is that fair?

These things would certainly be good to note in cert-manager/trust though.

### What does "self-signed" mean? Is my CA self-signed?

Self-signed means exactly what it says; a certificate is self-signed if it is signed by its own private key.

Self-signed is a commonly confused term, however, and is very frequently misused to mean "not publicly trusted". We tend to use terms
like "private PKI" to denote the situation where an organization might have their own internal CA certificates which wouldn't
be trusted outside of the organization.

As an example, there are _many_ self-signed certificates in [Mozilla's CA Certificate Program](https://wiki.mozilla.org/CA), but

> all of those certificates would usually be described as "publicly trusted".

Reviewer 3: praise: thank you for giving the mozilla CA bundle example. This is great for newcomers!

Your certificate is self-signed only if it's signed with its own key.

### What's the difference between "root", "intermediate", and "leaf" certificates?

cert-manager uses the following definitions:

#### Root Certificates

Roots are self-signed certificates and almost always marked as CA certificates. They're usually not sent over the wire

> during a TLS handshake because they need to be explicitly trusted in order to be validated.

Reviewer 3: random question: why do we say "trust root"? I assume that we mean "the root of what we trust", but shouldn't we rather say "the trusted root" or "the root that we trust" instead?

Reviewer 2: I went digging around the man pages on my laptop and found

```
$ man update-ca-trust
...
       /etc/pki/tls/certs/ca-bundle.crt
           Classic filename, file contains a list of CA certificates trusted for TLS server authentication usage, in the simple BEGIN/END CERTIFICATE file format, without distrust information. This file is a symbolic link that
           refers to the consolidated output created by the update-ca-trust command.

       /etc/pki/tls/certs/ca-bundle.trust.crt
           Classic filename, file contains a list of CA certificates in the extended BEGIN/END TRUSTED CERTIFICATE file format, which includes trust (and/or distrust) flags specific to certificate usage. This file is a symbolic link
           that refers to the consolidated output created by the update-ca-trust command.

```

Which says something about a distinction between PEM labels "CERTIFICATE" and "TRUSTED CERTIFICATE" and that is described in this SO answer:
 * https://stackoverflow.com/a/5356351

Author: > I assume I should understand "trust root" as "trusted root", is that correct? The term "trust root" seems a bit weird
> I guess we could also say "root of the chain of trust" too.

Yeah I think it's largely used interchangeably and it's usually pretty clear what's meant.

Author: > Which says something about a distinction between PEM labels "CERTIFICATE" and "TRUSTED CERTIFICATE"

This is good to bear in mind for https://github.com/cert-manager/trust/ perhaps :thinking:

> Roots are sometimes defined as "CA certificates which are explicitly trusted" - which can include certificates which

Reviewer 3: suggestion: em dash 👼

```suggestion
Roots are sometimes defined as "CA certificates which are explicitly trusted"---which can include certificates which
```

aren't self-signed. cert-manager doesn't use this definition.

Changing trust stores to include new roots or remove old ones is a non-trivial task which can take months or years for publicly

> trusted roots. For this reason, roots are usually issued with very long lifetimes - often on the order of decades.

Reviewer 3: ```suggestion
trusted roots. For this reason, roots are usually issued with very long lifetimes—often on the order of decades.
```

> trusted roots. For this reason, roots are usually issued with very long lifetimes - often on the order of decades.

Reviewer 3: suiggestion: we could mention the case of the [Let's Encrypt DST root CA X3](https://letsencrypt.org/docs/dst-root-ca-x3-expiration-september-2021/) intermediate which expired in Sept 2021 and broke many old Android phones that were not being updated anymore; these phones cannot access some website due to this

A nice example of "long lifetime" that still broke things 😅

Author: I think Let's Encrypt's chains are super interesting and might be relevant somewhere in this doc because it's an example of an explicitly trusted _intermediate_ certificate in public trust stores.

I think I'd prefer not to add the example here; I think this issue shows the need for trust stores to be kept up to date, but that's kinda tangential to what this terminology guide is trying to achieve. That's more of an example for a TLS best-practices guide IMO.

#### Intermediate Certificates

Intermediates are CA certificates signed by another CA. Most intermediates will be signed by a root certificate, but it's
possible to construct longer chains where an intermediate can be signed by another intermediate.

> Intermediate certificates are usually issued with a much shorter lifetime than the CA which signed them. On the public
> internet, intermediate certificates are used on network-connected machines for day-to-day issuance so that the
> highly-valuable root certificates can remain entirely offline.

Reviewer 2: ```suggestion
Intermediate certificates are usually issued with a much shorter lifetime than the CA which signed them. On the Internet, intermediate certificates are used on network-connected machines for day-to-day issuance so that the
highly-valuable root certificates can remain entirely offline.
```

While intermediate certificates can also be explicitly trusted via addition to a trust store, they're usually validated
by "walking up" the chain and validating signatures until an explicitly trusted self-signed root certificate is found.

#### Leaf Certificates

> Leaf certificates are usually not CA certificates, and are used to represent a particular identity; on the public internet,
> leaf certificates usually identify a particular domain, such as `example.com`.
> 
> Leaf certificates are generally at the end of a chain of certificates, and must be sent along with one or more intermediates
> to create a chain which can be validated by verifying signatures up to a trusted root certificate.

Reviewer 2: ```suggestion
Leaf certificates are never CA certificates, and are used to represent a particular identity; on the public internet,
leaf certificates usually identify a particular domain, such as `example.com`.

Leaf certificates are always at the end of a chain of certificates, and must be sent along with one or more intermediates
to create a chain which can be validated by verifying signatures up to a trusted root certificate.
```
I referred to: https://en.wikipedia.org/wiki/Public_key_certificate#End-entity_or_leaf_certificate and I know Wikipedia is pretty unreliable but is there anywhere that describes a CA leaf certificate.

I mean does leaf / root have a different meaning when talking about a certificate bundle (e.g. tls.crt) as opposed to certificates used by clients and servers on the Internet. Is a leaf certificate, one without any children in a particular bundle. 
Is a root certificate  any self-signed certificate OR a certificate who's parent is not included in the same bundle?

Author: > I referred to: https://en.wikipedia.org/wiki/Public_key_certificate#End-entity_or_leaf_certificate and I know Wikipedia is pretty unreliable but is there anywhere that describes a CA leaf certificate.

> Is a leaf certificate, one without any children in a particular bundle.

I can't think of why anyone would have a CA leaf certificate, I just know it's _technically_ possible. Probably too far out there to be worth mentioning; I'll remove that bit.

I think you could make the case that such a thing could exist, but I don't know why anyone would do that really and I probably shouldn't have accounted for it in this page.
